feat: Cohere v2 Implementation

sajdakabir opened the pull request:
## Title

Cohere v2 Implementation

## Relevant issues

#6980 


## Pre-Submission checklist
<img width="622" alt="Screenshot 2025-04-08 at 11 48 29 PM" src="https://github.com/user-attachments/assets/a30e3a0b-feec-47e2-8110-ed940e0e516d" />

<img width="580" alt="Screenshot 2025-04-08 at 11 48 18 PM" src="https://github.com/user-attachments/assets/f17fc6c5-b7af-4102-b022-4f0ef7907806" />

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on (`make test-unit`)[https://docs.litellm.ai/docs/extras/contributing_code]
- [ ] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type


🆕 New Feature


## pr dec

this pr adds support for cohere's v2 chat api in lightllm. the implementation includes –

 1.  created transformation_v2.py file to handle the cohere v2  api format
2. added support for citations in both streaming and non-streaming responses
3. Implemented tests for all functionality including:
         - regular chat completion
         - streaming responses
         - citations handling
         - and function calling

**testing:**
	- added mocked tests for all functionality
	- included a skippable real API test for developers to verify with actual api calls


**how to test** 


```

# run all cohere v2 tests
python -m pytest tests/llm_translation/test_cohere_v2.py -v

or

source venv/bin/activate && export CO_API_KEY='your cohere api key' && python -m pytest tests/llm_translation/test_cohere_v2.py -v

# to test with a real API key (**skipped by default**)
# 1. remove the @pytest.mark.skip decorator from test_cohere_v2_real_api_call
# 2. Set your cohere api key in the environment variable CO_API_KEY


```

sajdakabir committed:
feat: cohore_v2 support

sajdakabir committed:
fix: linting

sajdakabir committed:
Merge branch 'BerriAI:main' into main

sajdakabir committed:
fix: lint error

krrishdholakia commented on litellm/llms/cohere/chat/transformation_v2.py:
```diff
@@ -0,0 +1,375 @@
+            messages=messages, model=model, llm_provider="cohere_chat"
+        )
+        optional_params["messages"] = cohere_messages
+        optional_params["model"] = model.split("/")[-1]  # Extract model name from model string
```
why is this necessary?

sajdakabir replied:
because cohere's api expects just the model name (e.g–command-r) in the request, not the full provider/model format lightllm uses internally (e.g– cohere_chat_v2/command-r). the split operation removes the provider prefix and keep the model as chohere expects

krrishdholakia commented on tests/llm_translation/test_cohere_v2.py:
```diff
@@ -0,0 +1,999 @@
+@pytest.mark.parametrize("stream", [True, False])
+@pytest.mark.flaky(retries=3, delay=1)
+@pytest.mark.asyncio
+async def test_chat_completion_cohere_v2_citations(stream):
```
since these are all mock - can we please move this inside `tests/litellm`

sajdakabir replied:
sure

krrishdholakia commented on tests/llm_translation/test_cohere_v2.py:
```diff
@@ -0,0 +1,999 @@
+                },
+            ]
+            response = await litellm.acompletion(
+                model="cohere_chat_v2/command-r",
```
since v2 might be used for other cohere endpoints too - e.g. embeddings, can we do a more generic `cohere_v2/` provider instead

this will make it reusable in different places

sajdakabir replied:
okay

krrishdholakia commented on litellm/main.py:
```diff
@@ -2108,6 +2108,45 @@ def completion(  # type: ignore # noqa: PLR0915
+            if extra_headers is not None:
+                headers.update(extra_headers)
+
+            response = base_llm_http_handler.completion(
```
why is `client` not passed here?

sajdakabir committed:
fix: moved the test file to tests/litellm/ .
added missing client paremeter in main.py.
changed the provider name from cohere_chat_v2 to cohere_v2

krrishdholakia commented on tests/litellm/test_cohere_v2.py:
```diff
@@ -0,0 +1,199 @@
+
+import pytest
+
+import litellm
```
@sajdakabir please follow the documented file structure pattern for this folder - https://github.com/BerriAI/litellm/tree/main/tests/litellm 

<img width="1354" alt="Screenshot 2025-04-09 at 3 27 50 PM" src="https://github.com/user-attachments/assets/74ee5e39-ecd9-47c8-985c-d870b4c19dc9" />

krrishdholakia replied:
this file should not be on the root, as there is no corresponding `litellm/cohere_v2.py`

sajdakabir committed:
fix: remove model split operation

sajdakabir committed:
fix: change cohere test file location and keep a generic name for cohere_v2

sajdakabir committed:
Merge branch 'BerriAI:main' into main

sajdakabir committed:
Merge branch 'main' of github.com:sajdakabir/litellm

sajdakabir committed:
merge to main

sajdakabir committed:
fix: pass litellm_params paramete

krrishdholakia commented on tests/litellm/llms/cohere/chat/test_transformation_v2.py:
```diff
@@ -0,0 +1,998 @@
+import io
+import os
+
+sys.path.insert(
```
this path is incorrect. it needs to be where litellm/ local dir is accessible

krrishdholakia commented on litellm/llms/cohere/chat/transformation_v2.py:
```diff
@@ -0,0 +1,376 @@
+            if param == "stop":
+                optional_params["stop_sequences"] = value
+            if param == "tools":
+                optional_params["tools"] = value
```
cohere supports tool calling the openai spec?

sajdakabir replied:
yes ,https://docs.cohere.com/v2/docs/compatibility-api#function-calling
<img width="828" alt="Screenshot 2025-04-11 at 9 31 52 PM" src="https://github.com/user-attachments/assets/e916de16-f644-47ab-a030-341de4d6f923" />

sajdakabir committed:
fix: file path

sajdakabir committed:
refa: skip test if COHORE_API_KEY is not available

sajdakabir committed:
fix: tools oarameter handle

sajdakabir commented:
i have fixed the code to properly handle tool format transformation. now instead of directly mapping the open ai tools parameter in **map_openai_params**, i'm letting the **transform_request** method handle the conversion using the **_construct_cohere_tool** and **_translate_openai_tool_to_cohere** methods.

@krrishdholakia

sajdakabir committed:
fix: linting error

krrishdholakia commented:
Please keep all optional param translation within `map_openai_params`

sajdakabir committed:
reaf:moved tools parameter handling to map_openai_params

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=9832) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=9832) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=9832) it.</sub>

juleskuehn commented:
Bump! Currently with LlamaIndex + LiteLLM, Cohere models do not have access to chat history at all. (If you have a workaround, please LMK) All other providers I've tried work correctly (OpenAI, Google, Anthropic, Groq, Cerebras, ...)

themrzmaster commented:
bump
